arch/xtensa/core/mpu.c

```diff
@@ -701,3 +707,419 @@ void xtensa_mpu_init(void)
+		/* Assume end of memory if wrap around. */
+		end_addr = 0xFFFFFFFFU;
+	} else {
+		end_addr = ROUND_UP(end_addr, XCHAL_MPU_ALIGN);
+	}
+
+	first_enabled_idx = find_first_enabled_entry(entries);
+
+	/* Find the slot where each address belongs to. */
+	entry_slot_s = (struct xtensa_mpu_entry *)
+		       check_addr_in_mpu_entries(entries, start_addr, first_enabled_idx,
+						 &exact_s, &idx_s);
+	entry_slot_e = (struct xtensa_mpu_entry *)
+		       check_addr_in_mpu_entries(entries, end_addr, first_enabled_idx,
+						 &exact_e, &idx_e);
+
+	/* NULL means the address is not in map at all. Bail. */
+	if ((entry_slot_s == NULL) || (entry_slot_e == NULL)) {
+		ret = -EINVAL;
+		goto out;
+	}
+
+	/*
+	 * If ending address does not match the slot address exactly,
+	 * we will need to consider this entry for permission checking.
+	 */
+	if (!exact_e) {
+		idx_e++;
+	}
+
+	for (int i = idx_s; i < idx_e; i++) {
+		uint8_t access_rights = xtensa_mpu_entry_access_rights_get(&entries[i]);
+
+		if (write) {
+			/* Need to check write permission. */
+			switch (access_rights) {
+			case XTENSA_MPU_ACCESS_P_WO_U_WO:
+				__fallthrough;
+			case XTENSA_MPU_ACCESS_P_RW_U_RWX:
+				__fallthrough;
+			case XTENSA_MPU_ACCESS_P_RW_U_RW:
+				__fallthrough;
+			case XTENSA_MPU_ACCESS_P_RWX_U_RWX:
+				/* These permissions are okay. */
+				break;
+			default:
+				ret = -EPERM;
+				goto out;
+			}
+		} else {
+			/* Only check read permission. */
+			switch (access_rights) {
+			case XTENSA_MPU_ACCESS_P_RW_U_RWX:
+				__fallthrough;
+			case XTENSA_MPU_ACCESS_P_RW_U_RO:
+				__fallthrough;
+			case XTENSA_MPU_ACCESS_P_RWX_U_RX:
+				__fallthrough;
+			case XTENSA_MPU_ACCESS_P_RO_U_RO:
+				__fallthrough;
+			case XTENSA_MPU_ACCESS_P_RX_U_RX:
+				__fallthrough;
+			case XTENSA_MPU_ACCESS_P_RW_U_RW:
+				__fallthrough;
+			case XTENSA_MPU_ACCESS_P_RWX_U_RWX:
+				/* These permissions are okay. */
+				break;
+			default:
+				ret = -EPERM;
+				goto out;
+			}
+		}
+	}
+
+out:
+	return ret;
+}
+
+void xtensa_user_stack_perms(struct k_thread *thread)
+{
+	int ret;
+
+	uintptr_t stack_end_addr = thread->stack_info.start + thread->stack_info.size;
+
+	if (stack_end_addr < thread->stack_info.start) {
+		/* Account for wrapping around back to 0. */
+		stack_end_addr = 0xFFFFFFFFU;
+	}
+
+	(void)memset((void *)thread->stack_info.start,
+		     (IS_ENABLED(CONFIG_INIT_STACKS)) ? 0xAA : 0x00,
+		     thread->stack_info.size - thread->stack_info.delta);
+
+	/* Add stack to new domain's MPU map. */
+	ret = mpu_map_region_add(thread->arch.mpu_map,
+				 thread->stack_info.start, stack_end_addr,
+				 XTENSA_MPU_ACCESS_P_RW_U_RW,
+				 CONFIG_XTENSA_MPU_DEFAULT_MEM_TYPE,
+				 NULL);
+
+	xtensa_mpu_map_write(thread->arch.mpu_map);
+
+	/* Probably this fails due to no more available slots in MPU map. */
+	ARG_UNUSED(ret);
+	__ASSERT_NO_MSG(ret == 0);
+}
+
+void xtensa_swap_update_mpu_map(struct k_thread *incoming)
+{
+	xtensa_mpu_map_write(incoming->arch.mpu_map);
```

Comment:
Performance nit: this C function only ever gets called out of assembly code.  Would be smaller and save a whole stack frame to make this a macro that extracts the mpu_map and calls it directly.  Alternatively just have xtensa_mpu_map_write() take a k_thread argument to begin with and extract the map there.

Reply:
Changed.